# Genome Assembly using MaSuRCa
# Illumina reads processing

## 1. Filtering
Quality metrics can be assessed and visualised with FastQC. NOTE; Masurca itself requires non-filtered Illumina reads; however, it's still important to understand the quality of your Illumina data; and filtered/processed Illumina reads can also be used in other programmes (e.g Jellyfish)
### 1.a Adapter removal
Adapters can be removed from Illumina reads using Trimmomatic (version 0.39).
For this, you will need to know the adapters used for your Illumina reads.
We used NexteraPE (PE ILLUMINACLIP:NexteraPE-PE.fa:2:30:10)
### 1.b Contamination removal
For contamination removal, we can use Kraken2. Kraken 2 is a fast and memory efficient tool for taxonomic assignment of metagenomics sequencing reads.
We can use this to detect contamination from arachae, bacterial, viral and human sequences based on the  MiniKraken2 v2 8GB database.

```bash
#!/bin/bash
R1=path/to/illumina/r1
R2=path/to/illumina/r2
DB=path/to/MiniKraken2v28GBdatabase
kraken2 --db $databasePath --paired --gzip-compressed --unclassified-out useqs#.fq --classified-out cseqs#.fq --report kraken_report $R1 $R2
````

### 2 - Jellyfish
We can use a programmed called Jellyfish to estimate genome size.
This is via a k-mer frequency analyses.
We will use kmers of 17, 21 and 27. We can generate histograms for these kmers using the command histo and visualise them with GenomeScope.

```bash
#!/bin/bash

# Configuration Variables
KMER=17
ESTIMATED_SIZE=650000000
THREADS=12
INPUT_DIR=
OUTPUT_DIR=
# Input Files
R1=$INPUT_DIR/useqs_1_renamed.fq
R2=$INPUT_DIR/useqs_2_renamed.fq
CONCATENATED_FASTQ=$OUTPUT_DIR/illumina_kraken.catR1R2.fastq.gz

# Jellyfish Commands
# This section runs Jellyfish to count kmers and generate a histogram.
# It uses the concatenated FASTQ file as input.
jellyfish count -C -m $KMER -s $ESTIMATED_SIZE -t $THREADS $CONCATENATED_FASTQ -o $OUTPUT_DIR/illumina.catR1R2.${KMER}mer_counts.jf
jellyfish histo -t $THREADS $OUTPUT_DIR/illumina.catR1R2.${KMER}mer_counts.jf > $OUTPUT_DIR/illumina.catR1R2.${KMER}mer_counts.histo

# End of Script
````
