# Annotation
## Using Braker3 to annotate gene structures

Now that we have our masked genome, we can move onto genome annotation.
For this, we will use Braker3 
```bash 
('singularity build braker3.sif docker://teambraker/braker3:latest)
````
and annotate the genome via protein homology (using the vertebrate genome from  "https://bioinf.uni-greifswald.de/bioinf/partitioned_odb11/Vertebrata.fa.gz"). BRAKER predicts genes in a genomic sequence, identifying introns, exons, start and stop codons, and other features that define protein-coding genes; to this end, it outputs a set of predicted gene models. However, it does not functionally annotate these gene structures. That step comes after!

Note: we need to use the softmasked genome. Softmasking preserves the sequence information of repetitive regions while marking them as less significant. This allows gene prediction tools to still recognize and potentially use these regions if they contain legitimate biological signals that might be otherwise masked out.

### Run Braker3
```bash

# Load necessary modules
module load singularity
module load samtools/1.10

# Define paths and variables
GENOME_FA=/path/to/softmasked.fasta
PROT_SEQ=/path/to/braker_proteinDB_only/as_example/Vertebrata.fa
THREADS=30
BRAKER_SIF=$BRAKER_DIR/braker3.sif
WORKING_DIR=/path/to/working_dir
AUGUSTUS_CONFIG_PATH=/path/to/augustus/config

# Run BRAKER with the specified parameters
singularity exec -B ${PWD}:${PWD} \
    $BRAKER_SIF braker.pl \
    --genome=$GENOME_FA \
    --prot_seq=$PROT_SEQ \
    --threads=$THREADS \
    --workingdir=$WORKING_DIR \
    --AUGUSTUS_CONFIG_PATH=$AUGUSTUS_CONFIG_PATH \
    --gff3
````


## Functionally annotating gene structures.
Now that we have identified the gene structures, we can functionally annotate them. This means assigning biological information to the gene models, such as gene names, gene ontology terms, and pathway information. We will use the blastp and interproscan tools to do this.

### Blastp
Blastp is a tool that compares a protein query sequence against a protein sequence database. It identifies homologous sequences in the database and provides information about the function of the query sequence. We will use blastp to search for homologous sequences in the Swiss-Prot database (https://www.uniprot.org/uniprotkb)

```bash
# Load necessary modules
module load ncbi/blast+/2.6.0

WORKING_DIR=/path/to/working_dir
BRAKER=/path/to/braker_output/braker.aa
SWISSPROT_DB=path/to/swissprot_db

#need to do only once per database
gunzip -c $SWISSPROT_DB | makeblastdb -in - -dbtype prot -out swissprot.nb -title nr
echo "database built"

blastp -db $SWISSPROT_DB/nr.db -query $BRAKER \
-evalue 1e-6 -max_hsps 1 -max_target_seqs 1 -outfmt 6 -out blast_nr.out

##Explanation;
#An E-value of 1e-6 means that there is a 1 in 1,000,000 chance that the alignment is due to random chance.
#HSPs; High-scoring segment pairs; hsps 1 = show only highest match of query-database pairs 
````

### InterProScan
InterProScan is a tool that scans protein sequences against multiple databases to identify conserved domains, motifs, and functional sites. We will use InterProScan to functionally annotate the gene models with information about protein domains and functional sites.

```bash
INTERPROSCAN=/path/to/interproscan-5.68-100.0
BRAKER=/path/to/braker_output/braker.aa

$INTERPROSCAN/interproscan.sh -appl pfam -dp -f TSV -goterms -iprlookup -pa -t p -cpu \
 -i $BRAKER -o output.iprscan

 ````
### Integrating Blastp and Interproscan results
Next we will integrate both the Blastp and Interproscan results together into one annotation file.  For this, we will use AGAT (https://github.com/NBISweden/AGAT). AGAT has a multitude of tools that allows us to manage and manipulate GFF3 files.

```bash
BRAKER_GGF3=/path/to/braker_output/braker.gff3
BLAST_RESULTS=/path/to/blast_nr.out
BLAST_DB=/path/to/uniprot_sprot.fasta
INTERPROSCAN_RESULTS=/path/to/output.iprscan

agat_sp_manage_functional_annotation.pl -f $BRAKER_GFF3-b $BLAST_RESULTS --db $BLAST_DB -i $INTERPROSCAN_RESULTS --clean_name --clean_product --be 1e-6 -a 1e-6  -o braker_annotated.gff3
 ````
